src/dll.py

```
@@ -0,0 +1,104 @@
+        new_node = Node(val)
+        if len(self) == 0:
+            self.tail_node = new_node
+            self.head_node = new_node
+        else:
+            self.head_node.previous_node = new_node
+            new_node.next_node = self.head_node
+            self.head_node = new_node
+        self.length_list += 1
+        return self
+
+    def __len__(self):
+        """Return the length of a doubly linked list."""
+        return self.length_list
+
+    def append(self, val):
+        """Insert a new node to the tail of a doubly linked list."""
+        new_node = Node(val)
+        if len(self) == 0:
+            self.tail_node = new_node
+            self.head_node = new_node
+        else:
+            self.tail_node.next_node = new_node
+            new_node.previous_node = self.tail_node
+            self.tail_node = new_node
+        self.length_list += 1
+        return self
+
+    def pop(self):
+        """Remove a node from the head, return removed node value."""
+        if self.length_list == 0:
+            raise IndexError("can't pop off an empty list")
+        popped_node = self.head_node
+        self.head_node = self.head_node.next_node
+        if self.head_node:
+            self.head_node.previous_node = None
+        self.length_list -= 1
+        return popped_node.value
+
+    def shift(self):
+        """Remove a node from the tail, return removed node value."""
+        if self.length_list == 0:
+            raise IndexError("can't shift an empty list")
+        shifted_node = self.tail_node
+        self.tail_node = self.tail_node.previous_node
+        if self.tail_node:
+            self.tail_node.next_node = None
+        self.length_list -= 1
+        return shifted_node.value
+
+    def remove(self, val):
+        """Remove node with val from doubly linked list ."""
+        current_node = self.head_node
+        if not current_node:
+            raise IndexError("Can't remove from an empty list")
```

This is another situation where it is better to beg forgiveness than ask permission. The assignment specs say that trying to remove a value that is not in the list should raise an exception, so consider that you would get the same error if `self.head` was None at this point (i.e. the list is empty) and also if you reached the end of the list while looping without finding the `val`.